File: 140-word-break-ii/word-break-ii.py
class Solution:
    def wordBreak(self, s: str, wordDict: List[str]) -> List[str]:
        wordSet = set(wordDict)
        n = len(s)

        @lru_cache(maxsize=None)
        def wordBreakDP(start):
            if start == n:
                return [""]
            sentences = []
            for end in range(start + 1, n + 1):
                word = s[start:end]
                if word in wordSet:
                    next_sentences = wordBreakDP(end)
                    for sentence in next_sentences:
                        if sentence:
                            sentences.append(word + " " + sentence)
                        else:
                            sentences.append(word)
            return sentences

        return wordBreakDP(0)

File: 140-word-break-ii/word-break-ii.go
func wordBreak(s string, wordDict []string) []string {
    wordSet := make(map[string]bool)
    for _, word := range wordDict {
        wordSet[word] = true
    }

    var (
        sentences []string
        sentence  []string
        dfs       func(int)
    )

    dfs = func(start int) {
        if start == len(s) {
            sentences = append(sentences, strings.Join(sentence, " "))
            return
        }

        for end := start + 1; end <= len(s); end++ {
            word := s[start:end]
            if wordSet[word] {
                sentence = append(sentence, word)
                dfs(end)
                sentence = sentence[:len(sentence)-1]
            }
        }
    }

    dfs(0)
    return sentences
}
